### PR TITLE
Make uncaughtException from ThreadGroup subtypes VM-reachable

### PR DESCRIPTION
**Pull request by errt**
The `uncaughtException` method can be overridden in subtypes of ThreadGroup and overrides should be considered VM-reachable as per the JVM specification.

### CODE REVIEW

File: DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/threads/CreateThreadGroup.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package org.opalj.fpcf.fixtures.threads;
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/threads/CreateThreadGroup.java:2]: Add license declaration

File: DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/threads/CreateThreadGroup.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/threads/CreateThreadGroup.java]: Remove extra blank line

File: DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/threads/CreateThreadGroup.java [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+            line = 16,
+            resolvedTargets = {
+                    "Lorg/opalj/fpcf/fixtures/threads/MyRunnable;"})
+    public static void main(String[] args) {
+        ThreadGroup myThreadGroup = new MyThreadGroup();
+        Thread testThread = new Thread(myThreadGroup, new MyRunnable());
+        testThread.start();
+    }
+}
+class MyRunnable implements Runnable {
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/threads/CreateThreadGroup.java:19]: Add blank lines around class

File: DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/threads/CreateThreadGroup.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    @Override
+    public void run() {
+        System.out.println("Hello");
+    }
+}
+class MyThreadGroup extends ThreadGroup {
+
+
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/threads/CreateThreadGroup.java:27]: Remove extra blank line

File: DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/threads/CreateThreadGroup.java [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+import org.opalj.fpcf.properties.callgraph.DirectCall;
+import org.opalj.fpcf.properties.callgraph.VMReachable;
+
+public class CreateThreadGroup {
+    @DirectCall(
+            name = "run",
+            line = 16,
+            resolvedTargets = {
+                    "Lorg/opalj/fpcf/fixtures/threads/MyRunnable;"})
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/java/org/opalj/fpcf/fixtures/threads/CreateThreadGroup.java]: Fix formatting

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThreadRelatedCallsTest.scala [hunk 6]
```diff
@@ -0,0 +1,66 @@
+/* BSD 2-Clause License - see OPAL/LICENSE for details. */
+package org.opalj.fpcf
+
+import com.typesafe.config.{Config, ConfigValueFactory}
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThreadRelatedCallsTest.scala]: Remove multi-imports (and configure your IDE to not create them)

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThreadRelatedCallsTest.scala [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+import org.opalj.ai.domain.l1.DefaultReferenceValuesDomainWithCFGAndDefUse
+import org.opalj.ai.fpcf.properties.AIDomainFactoryKey
+import org.opalj.br.analyses.Project
+import org.opalj.br.analyses.cg.{InitialEntryPointsKey, InitialInstantiatedTypesKey}
+import org.opalj.tac.cg.{CHACallGraphKey, TypeIteratorKey}
+import org.opalj.tac.fpcf.analyses.LazyTACAIProvider
+import org.opalj.tac.fpcf.analyses.cg._
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThreadRelatedCallsTest.scala]: Remove underscore-import (and configure your IDE to not create them below 60 imports from the same package)

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThreadRelatedCallsTest.scala [hunk 8]
```diff
@@ -0,0 +1,66 @@
+/* BSD 2-Clause License - see OPAL/LICENSE for details. */
+package org.opalj.fpcf
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThreadRelatedCallsTest.scala]: Fix package declaration (we use `org.opal` on the first line, then one line for each subpackage)

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThreadRelatedCallsTest.scala [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+            case Some(requirements) => requirements ++ requiredDomains
+        }
+    }
+
+    override def createConfig(): Config = {
+        val baseConfig = super.createConfig()
+        // For these tests, we want to restrict entry points to "main" methods.
+        // Also, no types should be instantiated by default.
+        baseConfig.withValue(
+            InitialEntryPointsKey.ConfigKeyPrefix+"analysis",
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThreadRelatedCallsTest.scala:45]: Remove line break to improve formatting

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThreadRelatedCallsTest.scala [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+ * @author Julius Naeumann
+ */
+class ThreadRelatedCallsTest extends PropertiesTest {
+
+    override def init(p: Project[URL]): Unit = {
+        p.updateProjectInformationKeyInitializationData(TypeIteratorKey) {
+            case Some(_) => throw new IllegalArgumentException()
+            case None    => CHACallGraphKey.getTypeIterator(p)
+        }
+        val requiredDomains: Set[Class[_ <: AnyRef]] = Set(classOf[DefaultReferenceValuesDomainWithCFGAndDefUse[_]])
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThreadRelatedCallsTest.scala:33]: This line is probably too long (120 characters max.), so please break it

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/properties/callgraph/VMReachableMethodMatcher.scala [hunk 11]
```diff
@@ -0,0 +1,44 @@
+/* BSD 2-Clause License - see OPAL/LICENSE for details. */
+package org.opalj.fpcf.properties.callgraph
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/properties/callgraph/VMReachableMethodMatcher.scala]: Fix package declaration and multi-imports

File: OPAL/br/src/main/resources/reference.conf [hunk 12]
```diff
@@ -68,7 +68,7 @@ org.opalj {
             {declaringClass = "java/lang/Thread", name = "<init>", descriptor = "(Ljava/lang/ThreadGroup;Ljava/lang/String;)V"},
             {declaringClass = "java/lang/ThreadGroup", name = "<init>", descriptor = "()V"},
             {declaringClass = "java/lang/Thread", name = "exit", descriptor = "()V"},
-            {declaringClass = "java/lang/ThreadGroup", name = "uncaughtException", descriptor = "(Ljava/lang/Thread;Ljava/lang/Throwable;)V"},
+            {declaringClass = "java/lang/ThreadGroup+", name = "uncaughtException", descriptor = "(Ljava/lang/Thread;Ljava/lang/Throwable;)V"},
```
Comment (errt) [inline OPAL/br/src/main/resources/reference.conf]: Should not be necessary anymore

File: OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/ThreadRelatedCallsAnalysis.scala [hunk 13]
```diff
@@ -59,39 +59,42 @@ class ThreadStartAnalysis private[cg] (
         isDirect:        Boolean
     ): ProperPropertyComputationResult = {
         val indirectCalls = new IndirectCalls()
-
+        val vmReachableMethods = new VMReachableMethods()
```
Comment (errt) [inline OPAL/tac/src/main/scala/org/opalj/tac/fpcf/analyses/cg/ThreadRelatedCallsAnalysis.scala]: I think you should be able to just extend the indirect calls above with `with VMReachableMethodsBase, which might or might not be cleaner.`

File: DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThreadRelatedCallsTest.scala [hunk 14]
```diff
@@ -52,7 +55,8 @@ class ThreadRelatedCallsTest extends PropertiesTest {
     describe("ThreadStartAnalysis is executed") {
 
         val as = executeAnalyses(
-            Set(LazyTACAIProvider, AllocationSiteBasedPointsToAnalysisScheduler, CallGraphAnalysisScheduler, ThreadRelatedCallsAnalysisScheduler)
+            Set(LazyTACAIProvider, AllocationSiteBasedPointsToAnalysisScheduler,
```
Comment (errt) [inline DEVELOPING_OPAL/validate/src/test/scala/org/opalj/fpcf/ThreadRelatedCallsTest.scala]: Make a separate line for each one

File: OPAL/br/src/main/resources/reference.conf [hunk 15]
```diff
@@ -68,7 +68,6 @@ org.opalj {
             {declaringClass = "java/lang/Thread", name = "<init>", descriptor = "(Ljava/lang/ThreadGroup;Ljava/lang/String;)V"},
             {declaringClass = "java/lang/ThreadGroup", name = "<init>", descriptor = "()V"},
             {declaringClass = "java/lang/Thread", name = "exit", descriptor = "()V"},
-            {declaringClass = "java/lang/ThreadGroup+", name = "uncaughtException", descriptor = "(Ljava/lang/Thread;Ljava/lang/Throwable;)V"},
```
Comment (errt) [inline OPAL/br/src/main/resources/reference.conf]: Can we really remove it completely? This is for the standard uncaughtException. I think only the `+` for overrides should be removed

**Review by errt — COMMENTED** [review]
LGTM, someone please approve and merge

**Review by JuliusNmn — APPROVED** [review]

**Pull request merged**